src/main/java/org/jenkinsci/plugins/workflow/cps/CpsGroovyShell.java

```diff
@@ -78,4 +95,37 @@ protected synchronized String generateScriptName() {
         else
             return super.generateScriptName();
     }
+
+    private static class TimingLoader extends ClassLoader {
+        private final CpsFlowExecution execution;
+        TimingLoader(ClassLoader parent, CpsFlowExecution execution) {
+            super(parent);
+            this.execution = execution;
+        }
+        @Override protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+            execution.time(CpsFlowExecution.TimingKind.load, false);
+            try {
+                return super.loadClass(name, resolve);
+            } finally {
+                execution.time(CpsFlowExecution.TimingKind.load, true);
```

Comment:
Becomes a bit easier in `-source 8`, so long as each body returns at most one kind of checked exception.